allenact/embodiedai/aux_losses/losses.py

```diff
@@ -0,0 +1,599 @@
+    """
+
+    UUID = "InvDyn"
+
+    def __init__(
+        self, subsample_rate: float = 0.2, subsample_min_num: int = 10, *args, **kwargs
+    ):
+        """
+        Subsample the valid samples by the rate of `subsample_rate`,
+            if the total num of the valid samples is larger than `subsample_min_num`.
+        """
+        super().__init__(auxiliary_uuid=self.UUID, *args, **kwargs)
+
+        self.cross_entropy_loss = nn.CrossEntropyLoss(reduction="none")
+        self.subsample_rate = subsample_rate
+        self.subsample_min_num = subsample_min_num
+
+    def get_aux_loss(
+        self,
+        aux_model: nn.Module,
+        observations: ObservationType,
+        obs_embeds: torch.FloatTensor,
+        actions: torch.FloatTensor,
+        beliefs: torch.FloatTensor,
+        masks: torch.FloatTensor,
+        *args,
+        **kwargs
+    ):
+        ## we discard the last action in the batch
+        num_steps, num_sampler = actions.shape  # T, B
+        actions = cast(torch.LongTensor, actions)
+        actions = actions[:-1]  # (T-1, B)
+
+        ## find the final belief state based on masks
+        # we did not compute loss here as model.forward is compute-heavy
+        masks = masks.squeeze(-1)  # (T, B)
+
+        final_beliefs, _, _ = _propagate_final_beliefs_to_all_steps(
+            beliefs, masks, num_sampler, num_steps,
+        )
+
+        ## compute CE loss
+        decoder_in = torch.cat(
+            [obs_embeds[:-1], obs_embeds[1:], final_beliefs[:-1]], dim=2
+        )  # (T-1, B, *)
+
+        preds = aux_model(decoder_in)  # (T-1, B, A)
+        # cross entropy loss require class dim at 1
+        loss = self.cross_entropy_loss(
+            preds.view((num_steps - 1) * num_sampler, -1),  # ((T-1)*B, A)
+            actions.flatten(),  #  ((T-1)*B,)
+        )
+        loss = loss.view(num_steps - 1, num_sampler)  # (T-1, B)
+
+        def vanilla_valid_losses(loss, num_sampler, end_locs_batch):
+            ##  this is just used to verify the vectorized version works correctly.
+            ##  not used for experimentation
+            valid_losses = []
+            for i in range(num_sampler):
+                end_locs = end_locs_batch[i]
+                for j in range(len(end_locs)):
+                    if j == 0:
+                        start_loc = 0
+                    else:
+                        start_loc = end_locs[j - 1] + 1
+                    end_loc = end_locs[j]
+                    if end_loc - start_loc <= 0:  # the episode only 1-step
+                        continue
+                    valid_losses.append(loss[start_loc:end_loc, i])
+
+            if len(valid_losses) == 0:
+                valid_losses = torch.zeros(1, dtype=torch.float).to(loss)
+            else:
+                valid_losses = torch.cat(valid_losses)  # (sum m, )
+            return valid_losses
```

Comment:
Can you comment this out to make LGTM happy? Nice to keep it here for debugging.